### PR TITLE
Don't send unexposed classes to inspector plugins

### LINKED ISSUE #36078: InspectorPlugin.can_handle(arg) can receive objects which are not exposed to GdScript
**Godot version:**
40d1b0bfdb62d24d72f1f0102a7caf6f1c14e595

**OS/device including version:**
Ubuntu 18.04 64bit

**Issue description:**
When we create InspectorPlugin and we implement `can_handle(arg_obj)` function, I would expect the argument will be an Object (like its mentioned in docs). But it seems it's not always a case, in my particular case it was an instance of SectionedInspectorFilter (class which is not exposed to GdScript in the first place). I'm not 100% sure but I think this is an undesired behavior. 

This makes such implementations of `can_handle()` improper:
```
func can_handle(in_object):
	return "some_property" in in_object
```
If we will use that concrete implementation, then when navigating inside of Editor Settings, we will experience a flood of such warnings:
`core/project_settings.cpp:209 - Property not found: application/config/some_property`

inspector_plugin_can_handle

**Steps to reproduce:**
Short way:
1. Download reproduction project
2. Open Editor settings, notice there is a flood of `core/project_settings.cpp:209 - Property not found: application/config/some_property` warnings in the editor output
3. It can be fixed/worked around by correcting res://addons/com.plugin.issue/InspectorPlugin.gd `can_handle(arg)` implementation

Long way (general steps):
1. Create new Plugin and inspector plugin.
2. Inside InspectorPlugin implement can_handle method:
```
func can_handle(in_object):
	return "some_property" in in_object
```
3. Register inspector plugin inside the Plugin
4. Open Editor settings and take a look at console output, there should be a lot of `core/project_settings.cpp:209 - Property not found: application/config/some_property` errors

**Minimal reproduction project:**
[inspector_plugin_can_handle_issue.zip](https://github.com/godotengine/godot/files/4181191/inspector_plugin_can_handle_issue.zip)

### PR DESCRIPTION
**Pull request by KoBeWi**
Fixes #36078

### CODE REVIEW

**Comment by dsnopek** [comment]
I don't know that this is the right solution. I'm pretty sure it's always been possible to encounter objects with unexposed classes in GDScript, but in the past people have mostly not noticed, and the objects basically get treated as the closest base class that is exposed. I think the real problem is the error spam.

**Comment by KoBeWi** [comment]
The error spam seems fixed, there is only one warning printed now. There is no way to get rid of it; it's actually intended, because `is` operator will call `_get()` and ProjectSettings will print warning if the property does not exist.

That said, this indeed handles only one kind of plugins, so there might be a similar problems somewhere else. Not worth fixing probably, I just want to close the issue.

**Comment by dsnopek** [comment]
I'm just worried this will end up excluding something that a plugin does actually want to handle. There's lots of classes in the editor that aren't exposed, and both GDScript and GDExtension can interact with them just fine, so long as they treat them as their nearest exposed base class (there's code that specifically does that in godot-cpp).

**Comment by YuriSizov** [comment]
> (there's code that specifically does that in godot-cpp).

@dsnopek Could you point me towards that code? That would help validate the use case. Although I would say that you are not expected to do that. Maybe we should just cast the object to the nearest exposed class (somehow) before sending it to plugins.

**Comment by dsnopek** [comment]
> @dsnopek Could you point me towards that code?

I don't know how helpful it'll be, but it's in godot-cpp's `ClassDB::get_instance_binding_callbacks()`.

> Although I would say that you are not expected to do that.

I'm not sure what you mean by this? In GDExtension, we only generate code for exposed classes, so if we encounter an object from an unexposed class, we need to wrap it in something. Even if we avoid passing unexposed classes to the specific method in this PR, you can still encounter them, especially in the editor (just walking the scene tree would find some).

**Comment by YuriSizov** [comment]
> I'm not sure what you mean by this?

From your comment I assumed you want to intentionally handle unexposed classes:

> I'm just worried this will end up excluding something that a plugin does actually want to handle.

**Comment by dsnopek** [comment]
> From your comment I assumed you want to intentionally handle unexposed classes:
> 
> > I'm just worried this will end up excluding something that a plugin does actually want to handle.

I guess my point was just that there's unexposed classes all over the place in the editor, and in both GDScript and GDExtension (and probably C# too?) they'll basically get treated as their nearest exposed parent class, so hiding them could maybe lead to missing something you want.

tl;dr encountering unexposed classes is normal :-)

Looking at the issue this is meant to fix, it sounds like the problem here was a specific unexposed class (`EditorSettings`?) leading to error spam with a common pattern that's used in editor plugins. And it doesn't sound like the problem with the class is that it's unexposed, but the way it's `_get()` method works. So, it doesn't really seem like the problem is unexposed classes, but this particular class being passed to this particular function.

**Comment by dsnopek** [comment]
> Even though you could potentially get an unexposed class in other ways, you are not expected, from the userland, to be able to do much with it.
> ... [snip] ...
> On the one hand this means being more restrictive when it comes to hacking the editor, but on the other it's removing footguns.

IMO, interacting with unexposed classes isn't a footgun - they work fine!

As far as I understand it, the issue with the problematic class encountered here doesn't derive from it being unexposed. If you exposed it, the same problem would still happen.

**Comment by YuriSizov** [comment]
> IMO, interacting with unexposed classes isn't a footgun - they work fine!

Expecting them to work as their actual class is a footgun. Casting them to a more general and exposed type and using that is fine. But then my proposal to pre-cast them before passing a reference should be fine, I think.

If you expect to get an editor node or object and use its registered methods and signals, despite it not being exposed, you are playing with fire because we do not guarantee API compatibility for these methods and signals. They will work until they stop because we decided to rework some internal things.

**Comment by dsnopek** [comment]
> If you expect to get an editor node or object and use its registered methods and signals, despite it not being exposed, you are playing with fire because we do not guarantee API compatibility for these methods and signals. They will work until they stop because we decided to rework some internal things.

Ok, right, I agree with that!

However, as far as I understand the issue the PR aims to fix, that's not related to what's happening here.
